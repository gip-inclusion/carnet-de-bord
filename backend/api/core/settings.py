import os

from pydantic import BaseSettings

dir_path = os.path.dirname(os.path.realpath(__file__))


class Settings(BaseSettings):
    app_url: str
    database_url: str
    hasura_graphql_jwt_secret: str

    smtp_host: str
    smtp_port: str
    smtp_user: str | None
    smtp_pass: str | None

    min_pool_size: int = 2
    max_pool_size: int = 20

    PE_AUTH_BASE_URL: str = "https://entreprise.pole-emploi.fr"
    PE_BASE_URL: str = "https://api.emploi-store.fr"
    PE_SCOPE = "api_referentielagencesv1 organisationpe"
    PE_CLIENT_ID: str = "CLIENT_ID"
    PE_CLIENT_SECRET: str = "CLIENT_SECRET"

    V1_PREFIX: str = "/v1"
    MAIL_FROM: str = "support.carnet-de-bord@fabrique.social.gouv.fr"
    LOG_FORMAT = "[%(asctime)s:%(filename)s:%(lineno)s - %(funcName)20s() ] %(message)s"

    class Config:
        env_file = ".env"
        env_file_encoding = "utf-8"


# By default, load the file of the parent project if the ENV_FILE
# variable doesn't exist
#
# If you want to specify a specific env file, you can setup
# the ENV_FILE variable value
#
# Something like that: ENV_FILE="../.env" python scripts/connect_to_db.py

env_file_path = os.getenv(
    "ENV_FILE", os.path.join(dir_path, "..", "..", "..", "app", ".env")
)


if os.path.exists(env_file_path):
    settings = Settings(_env_file=env_file_path, _env_file_encoding="utf-8")
else:
    settings = Settings()
