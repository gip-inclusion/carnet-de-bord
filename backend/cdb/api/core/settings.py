import os

from pydantic import BaseSettings, validator

dir_path = os.path.dirname(os.path.realpath(__file__))


class Settings(BaseSettings):
    app_url: str
    database_url: str
    hasura_graphql_jwt_secret: str
    hasura_graphql_admin_secret: str
    graphql_api_url: str
    smtp_host: str
    smtp_port: str
    smtp_user: str | None
    smtp_pass: str | None

    min_pool_size: int = 2
    max_pool_size: int = 20

    PE_AUTH_BASE_URL: str = "https://entreprise.pole-emploi.fr"
    PE_BASE_URL: str = "https://api.pole-emploi.io"
    # PE_AUTH_BASE_URL: str = "https://entreprise.pe-qvr.fr"
    # PE_BASE_URL: str = "https://api.peio.pe-qvr.fr"
    PE_CLIENT_ID: str
    PE_CLIENT_SECRET: str

    action_secret: str

    V1_PREFIX: str = "/v1"
    MAIL_FROM: str = "contact@carnetdebord.inclusion.beta.gouv.fr"
    LOG_FORMAT = "[%(asctime)s:%(filename)s:%(lineno)s - %(funcName)20s() ] %(message)s"
    LOG_LEVEL: str = "INFO"
    LOG_AS_JSON: bool = True
    GQL_LOG_LEVEL: str = "WARNING"

    ENABLE_PEIO_API: bool = False
    ENABLE_SYNC_CONTRAINTES: bool = False

    SENTRY_DSN: str | None
    SENTRY_ENVIRONMENT: str | None

    @validator("LOG_LEVEL")
    def uppercase(raw: str) -> str:
        return raw.upper()

    class Config:
        env_file = ".env"
        env_file_encoding = "utf-8"


# By default, load the file of the parent project if the ENV_FILE
# variable doesn't exist
#
# If you want to specify a specific env file, you can setup
# the ENV_FILE variable value
#
# Something like that: ENV_FILE="../.env" python scripts/connect_to_db.py

env_file_path = os.getenv(
    "ENV_FILE", os.path.join(dir_path, "..", "..", "..", "..", ".env")
)


if os.path.exists(env_file_path):
    settings = Settings(_env_file=env_file_path, _env_file_encoding="utf-8")
else:
    settings = Settings()
