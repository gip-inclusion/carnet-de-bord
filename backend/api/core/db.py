import asyncpg
from asyncpg.pool import Pool

from api.core.settings import settings


async def get_connection_pool(database_url: str) -> Pool | None:
    return await asyncpg.create_pool(
        database_url, min_size=settings.min_pool_size, max_size=settings.max_pool_size
    )
