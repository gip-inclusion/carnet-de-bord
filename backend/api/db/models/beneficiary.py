from datetime import date, datetime
from uuid import UUID

from pydantic import BaseModel, Field

from api.db.models.notebook import Notebook


class Beneficiary(BaseModel):
    id: UUID
    email: str | None
    lastname: str
    firstname: str
    caf_number: str | None
    pe_number: str | None
    postal_code: str | None
    city: str | None
    address1: str | None
    address2: str | None
    mobile_number: str | None
    date_of_birth: date
    place_of_birth: str | None
    deployment_id: UUID
    created_at: datetime
    updated_at: datetime
    internal_id: str | None
    notebook: Notebook | None
    # BRSA users may not have an account
    # (account is created on the  first login attempt)
    account_id: UUID | None


class BeneficiaryImport(BaseModel):
    si_id: str = Field(..., title="Identifiant du SI interne")
    firstname: str = Field(..., title="Prénom")
    lastname: str = Field(..., title="Nom")
    date_of_birth: date = Field(..., title="Date de naissance")
    place_of_birth: str | None = Field(None, title="Lieu de naissance")
    phone_number: str | None = Field(None, title="Numéro de téléphone")
    email: str | None = Field(None, title="Email")
    address1: str | None = Field(None, title="Adresse")
    address2: str | None = Field(None, title="Complément d'adresse")
    postal_code: str | None = Field(None, title="Code postal")
    city: str | None = Field(None, title="Ville")
    work_situation: str | None = Field(None, title="Situation professionnelle")
    caf_number: str | None = Field(None, title="Numéro CAF/MSA")
    pe_number: str | None = Field(None, title="Identifiant Pole Emploi")
    right_rsa: str | None = Field(None, title="Droits RSA")
    right_are: bool | None = Field(None, title="Droits ARE")
    right_ass: bool | None = Field(None, title="Droits ASS")
    right_bonus: bool | None = Field(None, title="Droits Bonus")
    right_rqth: bool | None = Field(None, title="Droits RQTH")
    geographical_area: str | None = Field(None, title="Zone de mobilité")
    rome_code_description: str | None = Field(None, title="Emploi recherché")
    education_level: str | None = Field(None, title="Niveau de formation")
    structure_name: str | None = Field(None, title="Structure d'accompagnement")
    advisor_email: str | None = Field(None, title="Accompagnateur référent")
