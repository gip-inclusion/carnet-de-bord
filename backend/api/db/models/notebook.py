from datetime import datetime
from typing import List
from uuid import UUID

from pydantic import BaseModel
from strenum import StrEnum

from api.db.models.wanted_job import WantedJob


class Notebook(BaseModel):
    id: UUID
    beneficiary_id: UUID
    wanted_jobs: List[WantedJob]
    # @TODO: add other fields


class NotebookMemberInsert(BaseModel):
    notebook_id: UUID
    account_id: UUID
    last_visited_at: datetime | None = None
    member_type: str
    last_modified_at: datetime | None = None
    creator_id: datetime | None = None
    invitation_sent_at: datetime | None = None
    active: bool | None = None


class NotebookMember(NotebookMemberInsert):
    id: UUID
    created_at: datetime


class NotebookMemberTypeEnum(StrEnum):
    NO_REFERENT = "no_referent"
    REFERENT = "referent"
