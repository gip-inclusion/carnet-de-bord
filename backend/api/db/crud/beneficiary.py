from datetime import date
from typing import List
from uuid import UUID

from asyncpg import Record
from asyncpg.connection import Connection
from asyncpg.exceptions import UniqueViolationError

from api.db.crud.notebook import (
    NOTEBOOK_BASE_FIELDS,
    NOTEBOOK_BASE_JOINS,
    parse_notebook_from_record,
)
from api.db.models.beneficiary import Beneficiary, BeneficiaryImport

BENEFICIARY_BASE_QUERY = (
    """SELECT b.*, acc.id as account_id,"""
    + NOTEBOOK_BASE_FIELDS
    + """
    FROM public.beneficiary b
    LEFT JOIN notebook n ON n.beneficiary_id = b.id
    """
    + NOTEBOOK_BASE_JOINS
    + """
    LEFT JOIN public.account acc
    ON acc.beneficiary_id = b.id
"""
)


# see https://betakuang.medium.com/why-postgresqls-on-conflict-cannot-find-my-partial-unique-index-552327b85e1
firstname_lastname_date_of_birth_unique_idx = (
    "(LOWER(trim(firstname)), LOWER(trim(lastname)), date_of_birth, deployment_id)"
)


async def select_beneficiary(
    connection: Connection,
    beneficiary: BeneficiaryImport,
    deployment_id,
):
    return await connection.fetch(
        f"""
SELECT (firstname, lastname, date_of_birth, internal_id, deployment_id, id)
FROM beneficiary
WHERE (firstname = $1 AND lastname = $2 AND date_of_birth = $3)
OR (internal_id = $4 AND deployment_id = $5)
        """,
        beneficiary.firstname,
        beneficiary.lastname,
        beneficiary.date_of_birth,
        beneficiary.si_id,
        deployment_id,
    )


async def update_beneficiary(
    connection: Connection,
    beneficiary: BeneficiaryImport,
    deployment_id,
):
    await connection.fetchrow(
        f"""
UPDATE beneficiary SET mobile_number = $1
where internal_id = $2
returning id
        """,
        beneficiary.phone_number,
        beneficiary.si_id,
    )


async def insert_beneficiary(
    connection: Connection,
    beneficiary: BeneficiaryImport,
    deployment_id,
):
    await connection.fetchrow(
        f"""
INSERT INTO BENEFICIARY (firstname, lastname, internal_id, date_of_birth, deployment_id, mobile_number)
values ($1, $2, $3, $4, $5, $6)
returning id
        """,
        beneficiary.firstname,
        beneficiary.lastname,
        beneficiary.si_id,
        beneficiary.date_of_birth,
        deployment_id,
        beneficiary.phone_number,
    )


async def import_beneficiary(
    connection: Connection,
    beneficiary: BeneficiaryImport,
    deployment_id,
):
    result = select_beneficiary(connection, beneficiary, deployment_id)
    existing_rows = await result
    if len(existing_rows) == 0:
        print("inserting new user")
        await insert_beneficiary(connection, beneficiary, deployment_id)
    else:
        print(existing_rows[0][0][4])
        if (
            len(existing_rows) == 1
            and existing_rows[0][0][0] == beneficiary.firstname
            and existing_rows[0][0][1] == beneficiary.lastname
            and existing_rows[0][0][2] == beneficiary.date_of_birth
            and existing_rows[0][0][3] == beneficiary.si_id
        ):
            # Doesn't work somehow
            #     and existing_rows[0][0][4] == deployment_id
            print("updating existing user")
            await update_beneficiary(connection, beneficiary, deployment_id)
        else:
            print("data problem")


async def get_beneficiary_with_query(
    connection: Connection, query: str, *args
) -> Beneficiary | None:

    async with connection.transaction():

        beneficiary = None

        beneficiary_records: List[Record] = await connection.fetch(
            BENEFICIARY_BASE_QUERY + query,
            *args,
        )

        for beneficiary_record in beneficiary_records:

            if beneficiary is None:
                beneficiary = Beneficiary.parse_obj(beneficiary_record)

            if beneficiary_record["n_id"] is not None:
                beneficiary.notebook = await parse_notebook_from_record(
                    beneficiary_record,
                    notebook=beneficiary.notebook,
                )

        return beneficiary


async def get_beneficiary_from_personal_information(
    connection: Connection, firstname: str, lastname: str, birth_date: date
) -> Beneficiary | None:

    return await get_beneficiary_with_query(
        connection,
        # We use LOWER(BTRIM(field)) here to match the index so postgres
        # can use the index rather than scanning the whole table
        "WHERE LOWER(BTRIM(b.firstname)) = LOWER($1) AND "
        "LOWER(BTRIM(b.lastname)) = LOWER($2) AND "
        "b.date_of_birth = $3",
        firstname,
        lastname,
        birth_date,
    )


async def get_beneficiary_by_id(
    connection: Connection, beneficiary_id: UUID
) -> Beneficiary | None:

    return await get_beneficiary_with_query(
        connection, "WHERE b.id = $1", beneficiary_id
    )
