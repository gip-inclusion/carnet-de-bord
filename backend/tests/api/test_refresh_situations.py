from datetime import datetime, timedelta, timezone
from unittest import mock
from urllib.parse import quote
from uuid import UUID

import httpx
import pytest
import respx
from asyncpg import Connection, Record
from gql import gql

from cdb.api.core.settings import Settings, settings
from cdb.api.db.crud.external_data import (
    insert_external_data,
    insert_external_data_info,
)
from cdb.api.db.graphql import get_client
from cdb.api.db.models.beneficiary import Beneficiary
from cdb.api.db.models.external_data import (
    ExternalDataInfoInsert,
    ExternalDataInsert,
    ExternalSource,
)
from cdb.api.db.models.notebook import Notebook
from tests.mocks.pole_emploi_diagnostic import (
    PE_API_RECHERCHE_USAGERS_RESULT_OK_MOCK,
)
from tests.utils.approvaltests import verify_as_json

pytestmark = pytest.mark.graphql


@pytest.fixture
async def pe_settings():
    settings.PE_CLIENT_SECRET = "pe_client_secret"
    settings.PE_CLIENT_ID = "pe_client_id"

    yield settings


async def test_produces_a_401_when_the_token_is_missing(
    test_client: httpx.AsyncClient,
    beneficiary_sophie_tifour: Beneficiary,
):
    response = await test_client.post(
        "/v1/notebooks/refresh-situations-from-pole-emploi",
        json=build_payload(notebook_id=beneficiary_sophie_tifour.notebook.id),
    )
    assert response.status_code == 401
    json = response.json()
    assert json["detail"] == "Missing credentials"


@pytest.mark.graphql
@respx.mock
async def test_nominal(
    test_client: httpx.AsyncClient,
    beneficiary_sophie_tifour: Beneficiary,
    notebook_sophie_tifour: Notebook,
    sophie_tifour_contraintes: dict,
    pe_settings,
):
    mock_pe_api(
        settings,
        beneficiary_sophie_tifour,
        sophie_tifour_contraintes,
    )

    response = await call_refresh_api(beneficiary_sophie_tifour, test_client)

    expect_data_to_have_been_refreshed(True, response)

    situations = await get_notebook_situations_by_id(notebook_sophie_tifour.id)
    assert situations

    assert [
        (
            situation["refSituation"]["description"],
            situation["createdAt"].split("T")[0],
        )
        for situation in situations
    ] == [
        ("Doit quitter le logement", "2023-05-12"),
        ("Hébergé chez un tiers", "2021-09-21"),
    ]


@pytest.mark.graphql
@respx.mock
async def test_refreshes_on_new_data_from_pe(
    test_client: httpx.AsyncClient,
    beneficiary_sophie_tifour: Beneficiary,
    sophie_tifour_add_only_contraintes: dict,
    pe_settings,
):
    mock_pe_api(
        pe_settings,
        beneficiary_sophie_tifour,
        sophie_tifour_add_only_contraintes,
    )

    response = await call_refresh_api(beneficiary_sophie_tifour, test_client)

    expect_data_to_have_been_refreshed(True, response)


@pytest.mark.graphql
@respx.mock
@mock.patch("cdb.api.v1.routers.refresh_situations.call_pe")
async def test_gets_data_from_pe_when_its_missing(
    call_pe: mock.MagicMock,
    test_client: httpx.AsyncClient,
    beneficiary_sophie_tifour: Beneficiary,
):
    given_pe_data_is_different(True, call_pe)

    response = await call_refresh_api(beneficiary_sophie_tifour, test_client)

    expect_data_to_have_been_refreshed(True, response)


@pytest.mark.graphql
@respx.mock
@mock.patch("cdb.api.v1.routers.refresh_situations.call_pe")
async def test_does_nothing_when_pe_data_are_fresh(
    call_pe: mock.MagicMock,
    test_client: httpx.AsyncClient,
    beneficiary_sophie_tifour: Beneficiary,
    db_connection: Connection,
):
    external_data = await insert_external_data(
        db_connection,
        ExternalDataInsert(data={}, hash="hash", source=ExternalSource.PE_IO),
    )
    assert external_data
    await insert_external_data_info(
        db_connection,
        ExternalDataInfoInsert(
            external_data_id=external_data.id,
            beneficiary_id=beneficiary_sophie_tifour.id,
            created_at=datetime.now(tz=timezone.utc) - timedelta(minutes=59),
        ),
    )

    response = await call_refresh_api(beneficiary_sophie_tifour, test_client)

    assert call_pe.assert_not_called

    expect_data_to_have_been_refreshed(False, response)


@pytest.mark.graphql
@mock.patch("cdb.api.v1.routers.refresh_situations.call_pe")
async def test_refreshes_data_from_pe_when_they_have_expired(
    call_pe: mock.MagicMock,
    test_client: httpx.AsyncClient,
    beneficiary_sophie_tifour: Beneficiary,
    db_connection: Connection,
):
    given_pe_data_is_different(True, call_pe)
    await given_beneficiary_has_external_data_at_dates(
        [expired_date()], beneficiary_sophie_tifour, db_connection
    )

    response = await call_refresh_api(beneficiary_sophie_tifour, test_client)

    expect_data_to_have_been_refreshed(True, response)


@pytest.mark.graphql
@mock.patch("cdb.api.v1.routers.refresh_situations.call_pe")
async def test_does_nothing_when_at_least_one_data_is_fresh(
    call_pe: mock.MagicMock,
    test_client: httpx.AsyncClient,
    beneficiary_sophie_tifour: Beneficiary,
    db_connection: Connection,
):
    given_pe_data_is_different(True, call_pe)
    await given_beneficiary_has_external_data_at_dates(
        [fresh_creation_date(), expired_date()],
        beneficiary_sophie_tifour,
        db_connection,
    )

    response = await call_refresh_api(beneficiary_sophie_tifour, test_client)

    expect_data_to_have_been_refreshed(False, response)


@pytest.mark.graphql
@respx.mock
@mock.patch("cdb.api.v1.routers.refresh_situations.call_pe")
async def test_does_nothing_when_feature_is_disabled(
    call_pe: mock.MagicMock,
    test_client: httpx.AsyncClient,
    beneficiary_sophie_tifour: Beneficiary,
):
    settings.ENABLE_SITUATION_API = False

    response = await call_refresh_api(beneficiary_sophie_tifour, test_client)

    assert call_pe.assert_not_called
    expect_data_to_have_been_refreshed(False, response)

    settings.ENABLE_SITUATION_API = True


@pytest.mark.graphql
@respx.mock
async def test_saves_the_new_contraintes_into_external_data(
    test_client: httpx.AsyncClient,
    beneficiary_sophie_tifour: Beneficiary,
    sophie_tifour_add_only_contraintes: dict,
    db_connection: Connection,
    pe_settings,
):
    mock_pe_api(
        pe_settings, beneficiary_sophie_tifour, sophie_tifour_add_only_contraintes
    )
    await given_beneficiary_has_external_data_at_dates(
        [], beneficiary_sophie_tifour, db_connection
    )

    response = await call_refresh_api(beneficiary_sophie_tifour, test_client)

    record: Record = await get_last_saved_external_data(
        beneficiary_sophie_tifour.id, ExternalSource.PE_IO, db_connection
    )
    verify_as_json(
        {
            "data": record["data"],
            "source": record["source"],
            "hash": record["hash"],
        }
    )
    expect_data_to_have_been_refreshed(True, response)


@pytest.mark.graphql
@respx.mock
async def test_saves_the_new_contraintes_into_external_data_only_when_changed(
    test_client: httpx.AsyncClient,
    beneficiary_sophie_tifour: Beneficiary,
    notebook_sophie_tifour: Notebook,
    sophie_tifour_contraintes: dict,
    situation_id_contrainte_horaire: UUID,
    db_connection: Connection,
    pe_settings,
):
    mock_pe_api(pe_settings, beneficiary_sophie_tifour, sophie_tifour_contraintes)
    before = expired_date()

    await call_refresh_api(beneficiary_sophie_tifour, test_client)
    # update created_at so it happened in the past
    await update_external_data_created_at(
        beneficiary_sophie_tifour.id, before, db_connection
    )
    # add a new situation locally so there will be some differences in situations
    await add_situation_to_notebook(
        notebook_sophie_tifour.id, situation_id_contrainte_horaire, db_connection
    )

    response = await call_refresh_api(beneficiary_sophie_tifour, test_client)

    result = await get_last_saved_external_data(
        beneficiary_sophie_tifour.id, ExternalSource.PE_IO, db_connection
    )
    assert result["created_at"] == before
    expect_data_to_have_been_refreshed(False, response)


async def get_last_saved_external_data(
    beneficiary__id: UUID, source: ExternalSource, db_connection
) -> Record:
    result = await db_connection.fetchrow(
        """
        SELECT data, source, hash, edi.created_at
        FROM external_data
        LEFT JOIN external_data_info edi ON external_data_id = id
        WHERE beneficiary_id = $1 AND source = $2
        ORDER BY edi.created_at DESC
        """,
        beneficiary__id,
        source,
    )
    return result


async def update_external_data_created_at(
    beneficiary_id: UUID, created_at: datetime, db_connection
):
    await db_connection.fetchrow(
        """
        UPDATE external_data_info set created_at=$1
        WHERE beneficiary_id = $2
        """,
        created_at,
        beneficiary_id,
    )


async def add_situation_to_notebook(
    notebook_id: UUID, situation_id: UUID, db_connection
):
    await db_connection.fetchrow(
        """
        INSERT INTO notebook_situation(notebook_id, situation_id) VALUES($1, $2);
        """,
        notebook_id,
        situation_id,
    )


@pytest.mark.graphql
@respx.mock
async def test_does_nothing_when_data_from_pe_match_our_situations(
    test_client: httpx.AsyncClient,
    beneficiary_sophie_tifour: Beneficiary,
    sophie_tifour_no_new_contraintes: dict,
    pe_settings,
):
    mock_pe_api(
        pe_settings, beneficiary_sophie_tifour, sophie_tifour_no_new_contraintes
    )

    response = await call_refresh_api(beneficiary_sophie_tifour, test_client)

    expect_data_to_have_been_refreshed(False, response)


@pytest.mark.graphql
@respx.mock
async def test_get_no_situations(
    test_client: httpx.AsyncClient,
    beneficiary_sophie_tifour: Beneficiary,
    notebook_sophie_tifour: Notebook,
    empty_contraintes: dict,
    pe_settings,
):
    mock_pe_api(
        pe_settings,
        beneficiary_sophie_tifour,
        empty_contraintes,
    )

    response = await call_refresh_api(beneficiary_sophie_tifour, test_client)

    expect_data_to_have_been_refreshed(True, response)

    situations = await get_notebook_situations_by_id(notebook_sophie_tifour.id)
    assert len(situations) == 0


@pytest.mark.graphql
async def test_does_nothing_when_the_pe_client_id_is_missing(
    test_client: httpx.AsyncClient,
    notebook_sophie_tifour: Notebook,
    pe_settings,
):
    pe_settings.PE_CLIENT_ID = ""

    response = await test_client.post(
        "/v1/notebooks/refresh-situations-from-pole-emploi",
        headers={"secret-token": "action_secret_token"},
        json=build_payload(notebook_id=notebook_sophie_tifour.id),
    )

    expect_data_to_have_been_refreshed(False, response)


@pytest.mark.graphql
async def test_does_nothing_when_the_pe_client_secret_is_missing(
    test_client: httpx.AsyncClient,
    notebook_sophie_tifour: Notebook,
    pe_settings,
):
    pe_settings.PE_CLIENT_SECRET = ""

    response = await test_client.post(
        "/v1/notebooks/refresh-situations-from-pole-emploi",
        headers={"secret-token": "action_secret_token"},
        json=build_payload(notebook_id=notebook_sophie_tifour.id),
    )

    expect_data_to_have_been_refreshed(False, response)


@pytest.mark.graphql
@respx.mock
async def test_produces_a_400_when_the_pe_api_returns_a_500(
    test_client: httpx.AsyncClient,
    beneficiary_sophie_tifour: Beneficiary,
    pe_settings,
):
    respx.post(f"{pe_settings.PE_AUTH_BASE_URL}/connexion/oauth2/access_token").mock(
        return_value=httpx.Response(
            500,
            json={
                "token_type": "foo",
                "access_token": "batman",
                "expires_in": 3600,
            },
        )
    )
    response = await call_refresh_api(beneficiary_sophie_tifour, test_client)

    assert response.status_code == 400


def build_payload(notebook_id: UUID):
    return {
        "action": {"name": "refresh-situations-from-pole-emploi"},
        "request_query": "",
        "session_variables": {},
        "input": {
            "notebookId": notebook_id,
        },
    }


async def get_notebook_situations_by_id(id):
    async with get_client.gql_client_backend_only() as session:
        result = await session.execute(
            gql(
                """
            query situations($id: uuid!) {
                notebook_situation(
                    where: {notebookId: {_eq: $id}  deletedAt: {_is_null: true}}
                    order_by: {createdAt: desc}
                ) {
                    id, refSituation{ description} createdAt
                }
            }
        """
            ),
            variable_values={"id": id},
        )
        return result["notebook_situation"]


def mock_pe_api(
    settings: Settings,
    beneficiary: Beneficiary,
    contraintes: dict,
):
    respx.post(f"{settings.PE_AUTH_BASE_URL}/connexion/oauth2/access_token").mock(
        return_value=httpx.Response(
            200,
            json={
                "token_type": "foo",
                "access_token": "batman",
                "expires_in": 3600,
            },
        )
    )
    respx.post(
        f"{settings.PE_BASE_URL}/partenaire/rechercher-usager/v1/usagers/recherche",
        json={
            "nir": beneficiary.nir,
            "dateNaissance": beneficiary.date_of_birth.strftime("%Y-%m-%d"),
        },
    ).mock(
        return_value=httpx.Response(200, json=PE_API_RECHERCHE_USAGERS_RESULT_OK_MOCK)
    )

    pe_internal_id = PE_API_RECHERCHE_USAGERS_RESULT_OK_MOCK["identifiant"]

    respx.get(
        f"{settings.PE_BASE_URL}/partenaire/diagnosticargumente/v1/individus/{quote(pe_internal_id)}/contraintes",
    ).mock(return_value=httpx.Response(200, json=contraintes))


@pytest.fixture(scope="session")
def empty_contraintes() -> dict:
    return {
        "conseiller": "TNAN0260",
        "dateDeModification": "2023-05-12T12:54:39.000+00:00",
        "code": "7",
        "libelle": "Résoudre ses contraintes personnelles",
        "contraintes": [],
    }


@pytest.fixture(scope="session")
def sophie_tifour_contraintes() -> dict:
    return {
        "conseiller": "TNAN0260",
        "dateDeModification": "2023-05-12T12:54:39.000+00:00",
        "code": "7",
        "libelle": "Résoudre ses contraintes personnelles",
        "contraintes": [
            {
                "id": 27,
                "nom": "Faire face à des difficultés de logement",
                "valeur": "NON_ABORDEE",
                "date": "2023-05-12T12:54:39.000+00:00",
                "situations": [
                    {
                        "code": "24",
                        "libelle": "Hébergé chez un tiers",
                        "valeur": "OUI",
                    },
                    {
                        "code": "27",
                        "libelle": "Doit quitter le logement",
                        "valeur": "OUI",
                    },
                ],
                "objectifs": [],
            },
            {
                "id": 29,
                "nom": "Faire face à des difficultés administratives ou juridiques",
                "valeur": "CLOTUREE",
                "date": None,
                "situations": [
                    {
                        "code": "36",
                        "libelle": "Besoin d'etre guidé dans le cadre d'un accès aux droits",  # noqa: E501
                        "valeur": "CLOTUREE",
                    },
                ],
                "objectifs": [],
            },
        ],
    }


@pytest.fixture(scope="session")
def sophie_tifour_no_new_contraintes() -> dict:
    return {
        "conseiller": "TNAN0260",
        "dateDeModification": "2023-05-12T12:54:39.000+00:00",
        "code": "7",
        "libelle": "Résoudre ses contraintes personnelles",
        "contraintes": [
            {
                "id": 27,
                "nom": "Faire face à des difficultés de logement",
                "valeur": "NON_ABORDEE",
                "date": "2023-05-12T12:54:39.000+00:00",
                "situations": [
                    {
                        "code": "24",
                        "libelle": "Hébergé chez un tiers",
                        "valeur": "OUI",
                    },
                    {
                        "code": "27",
                        "libelle": "Besoin d'être guidé dans le cadre d'un accès aux droits",  # noqa: E501
                        "valeur": "OUI",
                    },
                    {
                        "code": "26",
                        "libelle": "Prêt  à suivre une formation",
                        "valeur": "OUI",
                    },
                ],
                "objectifs": [],
            },
        ],
    }


@pytest.fixture(scope="session")
def sophie_tifour_add_only_contraintes() -> dict:
    return {
        "conseiller": "TNAN0260",
        "dateDeModification": "2023-05-12T12:54:39.000+00:00",
        "code": "7",
        "libelle": "Résoudre ses contraintes personnelles",
        "contraintes": [
            {
                "id": 27,
                "nom": "Faire face à des difficultés de logement",
                "valeur": "NON_ABORDEE",
                "date": "2023-05-12T12:54:39.000+00:00",
                "situations": [
                    {
                        "code": "24",
                        "libelle": "Hébergé chez un tiers",
                        "valeur": "OUI",
                    },
                    {
                        "code": "27",
                        "libelle": "Besoin d'être guidé dans le cadre d'un accès aux droits",  # noqa: E501
                        "valeur": "OUI",
                    },
                    {
                        "code": "26",
                        "libelle": "Prêt  à suivre une formation",
                        "valeur": "OUI",
                    },
                ],
                "objectifs": [],
            },
            {
                "id": 29,
                "nom": "Faire face à des difficultés administratives ou juridiques",
                "valeur": "CLOTUREE",
                "date": "2022-05-12T12:54:39.000+00:00",
                "situations": [
                    {
                        "code": "36",
                        "libelle": "Attend un enfant ou plus",
                        "valeur": "OUI",
                    },
                ],
                "objectifs": [],
            },
        ],
    }


async def given_beneficiary_has_external_data_at_dates(
    dates, beneficiary, db_connection
):
    for date in dates:
        await given_beneficiary_has_pe_io_data(beneficiary, date, db_connection)


def expect_data_to_have_been_refreshed(refreshed, response):
    assert response.status_code == 200
    assert response.json() == {"data_has_been_updated": refreshed}


async def call_refresh_api(beneficiary, test_client):
    response = await test_client.post(
        "/v1/notebooks/refresh-situations-from-pole-emploi",
        headers={"secret-token": "action_secret_token"},
        json=build_payload(notebook_id=beneficiary.notebook.id),
    )
    return response


def fresh_creation_date():
    return expired_date() + timedelta(minutes=1)


async def given_beneficiary_has_pe_io_data(beneficiary, created_at, db_connection):
    external_data = await insert_external_data(
        db_connection,
        ExternalDataInsert(data={}, hash="hash", source=ExternalSource.PE_IO),
    )
    assert external_data
    await insert_external_data_info(
        db_connection,
        ExternalDataInfoInsert(
            external_data_id=external_data.id,
            beneficiary_id=beneficiary.id,
            created_at=created_at,
        ),
    )


async def given_external_data_with_pe_io(db_connection):
    external_data = await insert_external_data(
        db_connection,
        ExternalDataInsert(data={}, hash="hash", source=ExternalSource.PE_IO),
    )
    assert external_data
    return external_data


def given_pe_data_is_different(different, call_pe):
    call_pe.return_value = different


def expired_date():
    return datetime.now(tz=timezone.utc) - timedelta(hours=1)
