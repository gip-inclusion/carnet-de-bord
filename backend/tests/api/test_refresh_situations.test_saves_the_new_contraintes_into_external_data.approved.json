{
    "data": "[{\"id\": 27, \"nom\": \"Faire face \u00e0 des difficult\u00e9s de logement\", \"date\": \"2023-05-12T12:54:39+00:00\", \"valeur\": \"NON_ABORDEE\", \"situations\": [{\"code\": \"24\", \"valeur\": \"OUI\", \"libelle\": \"H\u00e9berg\u00e9 chez un tiers\"}, {\"code\": \"27\", \"valeur\": \"OUI\", \"libelle\": \"Besoin d'\u00eatre guid\u00e9 dans le cadre d'un acc\u00e8s aux droits\"}, {\"code\": \"26\", \"valeur\": \"OUI\", \"libelle\": \"Pr\u00eat  \u00e0 suivre une formation\"}]}, {\"id\": 29, \"nom\": \"Faire face \u00e0 des difficult\u00e9s administratives ou juridiques\", \"date\": \"2022-05-12T12:54:39+00:00\", \"valeur\": \"CLOTUREE\", \"situations\": [{\"code\": \"36\", \"valeur\": \"OUI\", \"libelle\": \"Attend un enfant ou plus\"}]}]",
    "hash": "14618249690d708291ade94cb220d1d28a51e0c5e52e996d90142a9e71eb14ba",
    "source": "peio"
}
